base_location/models/partner.py

```diff
@@ -2,17 +2,79 @@
 # Copyright 2016 Nicolas Bessi, Camptocamp SA
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
-from odoo import models, fields, api
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
 
 
 class ResPartner(models.Model):
     _inherit = 'res.partner'
-    zip_id = fields.Many2one('res.better.zip', 'City/Location')
+    zip_id = fields.Many2one('res.better.zip', 'ZIP Location')
+
+    @api.onchange('city_id')
+    def _onchange_city_id(self):
+        if not self.zip_id:
+            super(ResPartner, self)._onchange_city_id()
+        if self.zip_id and self.city_id != self.zip_id.city_id:
+            self.zip_id = False
+            self.zip = False
+            self.city = False
+        if self.city_id:
+            return {
+                'domain': {
+                    'zip_id': [('city_id', '=', self.city_id.id)]
+                },
+            }
+        return {'domain': {'zip_id': []}}
+
+    @api.onchange('state_id')
+    def _onchange_state_id(self):
```

Comment:
Is this really needed? Nobody searches zips only filtering the state: or you type the zip code or you type the city name

Reply:
This comment still remains

Reply:
It has been modified the domain filtering, but not the changes on fields (in order to avoid discrepancies)